python/src/array.cpp

```diff
@@ -773,6 +774,24 @@ void init_array(py::module_& m) {
             return a.shape(0);
           })
       .def("__iter__", [](const array& a) { return ArrayPythonIterator(a); })
+      .def(py::pickle(
+          [](array& a) { // __getstate__
+            if (a.dtype() == bfloat16) {
+              throw std::runtime_error("Pickle not supported for bfloat16.");
+            }
+            return py::array(buffer_info(a));
+          },
+          [](py::array npa) { // __setstate__
+            if (not py::isinstance<py::array>(npa)) {
+              throw std::runtime_error("Invalid state!");
+            }
+            return np_array_to_mlx(npa, std::nullopt);
+          }))
+      .def("__copy__", [](const array& self) { return array(self); })
+      .def(
+          "__deepcopy__",
+          [](const array& self, py::dict) { return array(self); },
+          "memo"_a)
```

Comment:
I'm not certain about having a `__deeepcopy__`. Could you say more about why you added it?

It's not a true deep copy so it might lead to unexpected/ambiguous behaviors.

For example:

```python
def fun(x):
  y = deepcopy(x)
  return y

# What should that be? 0 or 1?
dfdx = mx.grad(fun)(mx.array(1.0))
```

Reply:
I have added it because it speeds up the code (even if it is so fast that it should ring a bell, see my message https://github.com/ml-explore/mlx/pull/713#discussion_r1496130452), and I think that `copy` and `deepcopy` are the same thing for `mx.array`.

Regarding the behaviour in the example you suggested, here is how `jax` handles it compared to the current implementation
```python
def fun_deepcopy(x):
  y = deepcopy(x)
  return y
def fun_copy(x):
  y = copy(x)
  return y
print('mx copy: ', mx.grad(fun_copy)(mx.array(1.0)))
print('mx deepcopy: ', mx.grad(fun_deepcopy)(mx.array(1.0)))
print('jax copy: ', jgrad(fun_copy)(jnp.array(1.0)))
print('jax deepcopy: ', (fun_deepcopy)(jnp.array(1.0)))
```
Output:
```
mx copy:  array(1, dtype=float32)
mx deepcopy:  array(1, dtype=float32)
jax copy:  1.0
jax deepcopy:  1.0
```
If instead I remove the actual implementation of `__deepcopy__` (meaning that it is delegated to `__getstate__`+`__setstate__`), they do not match anymore
```
mx copy:  array(1, dtype=float32)
mx deepcopy:  array(0, dtype=float32)
jax copy:  1.0
jax deepcopy:  1.0
```

Now we can discuss what should be the right thing to return. Probably the `copy`s should refer to the whole object, computational graph included, so the correct answer is 1 in both the cases.  If one wants to copy just the values, it should use something like `detach()` (PyTorch) or `stop_gradient` (Jax), combined with some copy functions. 

See [this issue](https://github.com/google/jax/issues/2632) and the folliwing pull request to  see how they handled it in Jax.

Reply:
Thank you for the thoughtful response!! That is exactly what I was looking for!